### ISSUE TITLE
`explicit_auto_deref` suggestion leads to non compiling code

### ISSUE THREAD

**Issue by jonasbb** [rust-lang/rust-clippy#9101]
### Summary

The new `explicit_auto_deref` creates suggestions which do not compile. The lint already got ignored in serde <https://github.com/serde-rs/serde/commit/0ee71c70af8e0b6f29c1c696538fb2ebc081e2d2>.

### Lint Name

explicit_auto_deref

### Reproducer

I tried this code:

```rust
use serde::Serializer;

fn serialize<S, const N: usize>(bytes: &&[u8; N], serializer: S) -> Result<S::Ok, S::Error>
where
    S: Serializer,
{
    serializer.serialize_bytes(*bytes)
}
```
[Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=5eb1c6ab8a5702ff4b90b2600404db99)

I saw this happen:

```
warning: deref which would be done by auto-deref
 --> src/lib.rs:7:32
  |
7 |     serializer.serialize_bytes(*bytes)
  |                                ^^^^^^ help: try this: `bytes`
  |
  = note: `#[warn(clippy::explicit_auto_deref)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#explicit_auto_deref
```

I expected to see this happen:

No lint should be emitted, since otherwise the function does not compile.

```
error[[E0308]](https://doc.rust-lang.org/nightly/error-index.html#E0308): mismatched types
   --> src/lib.rs:7:32
    |
7   |     serializer.serialize_bytes(bytes)
    |                --------------- ^^^^^ expected slice `[u8]`, found `&[u8; N]`
    |                |
    |                arguments to this function are incorrect
    |
    = note: expected reference `&[u8]`
               found reference `&&[u8; N]`
note: associated function defined here
   --> /playground/.cargo/registry/src/github.com-1ecc6299db9ec823/serde-1.0.137/src/ser/mod.rs:737:8
    |
737 |     fn serialize_bytes(self, v: &[u8]) -> Result<Self::Ok, Self::Error>;
    |        ^^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0308`.
error: could not compile `playground` due to previous error
```

### Version

```text
Rust nightly clippy on the Playground.
clippy 0.1.64 (2022-07-01 46b8c23)

Build using the Nightly version: 1.64.0-nightly
(2022-07-01 46b8c23f3eb5e4d0e0aa)
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by tamaroning** [comment]
According to [RFC](https://github.com/rust-lang/rfcs/blob/73ee34aada32725bda1401dbd1f251e72f5d00d6/text/0401-coercions.md), I think clippy's suggestion is correct.
But rustc does not support all transitive type coercion for now (https://github.com/rust-lang/rust/issues/18602)

In code below, changing line 3 to `let _: &[u8] = a;` is fine because of auto deref and unsized coercion.
`&&[u8;2] -> &[u8;2] -> &[u8]`

```rust
fn main() {
    let a: &&[u8; 2] = &&[1, 2];
    let _: &[u8] = *a;
}
```
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=b14623210347777dcdcfdbf8ae73a16a
